server/app/controllers/api/CiviFormApiController.java

```diff
@@ -42,16 +48,24 @@ public Result checkAuth() {
   }
 
   protected void assertHasProgramReadPermission(Http.Request request, String programSlug) {
-    ApiKey apiKey =
-        profileUtils
-            .currentApiKey(request)
-            .orElseThrow(() -> new AccountNonexistentException("No API key found for profile"));
+    ApiKey apiKey = getApiKey(request);
 
     if (!apiKey.getGrants().hasProgramPermission(programSlug, ApiKeyGrants.Permission.READ)) {
       throw new UnauthorizedApiRequestException(apiKey, programSlug);
     }
   }
 
+  protected ApiKey getApiKey(Http.Request request) {
+    return profileUtils
+        .currentApiKey(request)
+        .orElseThrow(() -> new AccountNonexistentException("No API key found for profile"));
+  }
+
+  protected CompletionStage<ApiKey> recordApiKeyUsage(Http.Request request) {
+    return apiKeyRepository.recordApiKeyUsage(
```

Comment:
Blocking: This pattern requires each API endpoint to remember to call recordApiKeyUsage, which seems like it might be a source of potential errors as new endpoints are added. What about pushing the logic up to something that doesn't require each endpoint to remember to call it?

The current logic will throw if we can't record the API key usage, so pushing it up to more common logic shouldn't change any behavior around errors-for-failed-increments. The only thing we lose is a bit of asynchronous nature where the increment in parallel with the program definition lookup. For API calls, I assume this performance hit would be extremely minor.

Given the above, three potential options for where this logic could reside are:
1) ApiAuthenticator - We could make the increment call after we validate the cached key. Downside of putting it here is that if we wanted to track key usage by method, this wouldn't easily know what "route" the request was for.
2) CiviformApiController - The base class could have an abstract method for executing the request (e.g. `executeInternal`) and a final method `execute`. Implementors would put actual logic in `executeInternal` and the base class's `execute` would be responsible for incrementing the API key. This way we'd ensure logic was applied consistently. Additionally, it'd make it easier to add metrics on what route the API key was used for.
3) Refactor profileUtils.currentApiKey to increment.

Reply:
I ended up implementing it using a filter. No unit test but a browser test covers the behavior. PTAL and LMK what you think!